From 6501635a133fdf5b1f2ac17430d872efbafd84fa Mon Sep 17 00:00:00 2001
From: Nobuhiro Iwamatsu <iwamatsu@nigauri.org>
Date: Thu, 10 May 2012 05:45:12 +0900
Subject: [PATCH] support static library

Signed-off-by: Nobuhiro Iwamatsu <iwamatsu@nigauri.org>
---
 cmake/OpenCVModule.cmake |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/cmake/OpenCVModule.cmake b/cmake/OpenCVModule.cmake
index 9c21c73..4447b03 100644
--- a/cmake/OpenCVModule.cmake
+++ b/cmake/OpenCVModule.cmake
@@ -89,6 +89,7 @@ macro(ocv_add_module _name)
   string(TOLOWER "${_name}" name)
   string(REGEX REPLACE "^opencv_" "" ${name} "${name}")
   set(the_module opencv_${name})
+  set(the_module_s opencv_${name}_s)
 
   # the first pass - collect modules info, the second pass - create targets
   if(OPENCV_INITIAL_PASS)
@@ -432,6 +433,7 @@ endmacro()
 # ocv_create_module(<extra link dependencies>)
 macro(ocv_create_module)
   add_library(${the_module} ${OPENCV_MODULE_TYPE} ${OPENCV_MODULE_${the_module}_HEADERS} ${OPENCV_MODULE_${the_module}_SOURCES})
+  add_library(${the_module_s} STATIC ${OPENCV_MODULE_TYPE} ${OPENCV_MODULE_${the_module}_HEADERS} ${OPENCV_MODULE_${the_module}_SOURCES})
   target_link_libraries(${the_module} ${OPENCV_MODULE_${the_module}_DEPS} ${OPENCV_MODULE_${the_module}_DEPS_EXT} ${OPENCV_LINKER_LIBS} ${IPP_LIBS} ${ARGN})
   add_dependencies(opencv_modules ${the_module})
 
-- 
1.7.10

