--- a/modules/highgui/src/cap_v4l.cpp	2011-12-21 13:30:03.000000000 +0900
+++ b/modules/highgui/src/cap_v4l.cpp	2011-12-21 13:30:29.000000000 +0900
@@ -211,7 +211,6 @@
 #include <fcntl.h>
 #include <errno.h>
 #include <sys/ioctl.h>
-#include <sys/types.h>
 #include <sys/mman.h>
 
 #include <linux/videodev.h>
--- a/modules/highgui/src/cap_libv4l.cpp	2011-12-21 13:30:12.000000000 +0900
+++ b/modules/highgui/src/cap_libv4l.cpp	2011-12-21 13:30:52.000000000 +0900
@@ -232,7 +232,6 @@
 #include <unistd.h>
 #include <fcntl.h>
 #include <errno.h>
-#include <sys/types.h>
 #include <sys/mman.h>
 #include <string.h>
 #include <stdlib.h>
